spacy/pipeline/pipes.pyx

```diff
@@ -1063,51 +1064,205 @@ cdef class EntityRecognizer(Parser):
+
+    def set_kb(self, kb):
+        self.kb = kb
+
+    def require_model(self):
+        # Raise an error if the component's model is not initialized.
+        if getattr(self, "model", None) in (None, True, False):
+            raise ValueError(Errors.E109.format(name=self.name))
+
+    def require_kb(self):
+        # Raise an error if the knowledge base is not initialized.
+        if getattr(self, "kb", None) in (None, True, False):
+            raise ValueError(Errors.E139.format(name=self.name))
+
+    def begin_training(self, get_gold_tuples=lambda: [], pipeline=None, sgd=None, **kwargs):
+        self.require_kb()
+        self.cfg["entity_width"] = self.kb.entity_vector_length
+
+        if self.model is True:
+            self.model = self.Model(**self.cfg)
+
+        if sgd is None:
+            sgd = self.create_optimizer()
+        return sgd
+
+    def update(self, docs, golds, state=None, drop=0.0, sgd=None, losses=None):
+        self.require_model()
+        self.require_kb()
+
+        if losses is not None:
+            losses.setdefault(self.name, 0.0)
+
+        if not docs or not golds:
+            return 0
+
+        if len(docs) != len(golds):
+            raise ValueError(Errors.E077.format(value="EL training", n_docs=len(docs),
+                                                n_golds=len(golds)))
+
+        if isinstance(docs, Doc):
+            docs = [docs]
+            golds = [golds]
+
+        context_docs = list()
+        entity_encodings = list()
+
+        for doc, gold in zip(docs, golds):
+            for entity in gold.links:
+                start, end, gold_kb = entity
+                mention = doc.text[start:end]
+
+                candidates = self.kb.get_candidates(mention)
+                for c in candidates:
+                    kb_id = c.entity_
+                    # Currently only training on the positive instances
+                    if kb_id == gold_kb:
+                        prior_prob = c.prior_prob
+                        entity_encoding = c.entity_vector
+                        entity_encodings.append(entity_encoding)
+                        context_docs.append(doc)
+
+        if len(entity_encodings) > 0:
+            context_encodings, bp_context = self.model.begin_update(context_docs, drop=drop)
+            entity_encodings = np.asarray(entity_encodings, dtype=np.float32)
+
+            loss, d_scores = self.get_loss(scores=context_encodings, golds=entity_encodings, docs=None)
+            bp_context(d_scores, sgd=sgd)
+
+            if losses is not None:
+                losses[self.name] += loss
+            return loss
+        return 0
+
+    def get_loss(self, docs, golds, scores):
+        targets = [[1] for _  in golds]  # assuming we're only using positive examples
+        loss, gradients = self.get_cossim_loss_2(yh=scores, y=golds, t=targets)
+        loss = loss / len(golds)
+        return loss, gradients
+
+    def get_cossim_loss_2(self, yh, y, t):
```

Comment:
Is there a bug in the cosine loss function that `spacy pretrain` is using?

Reply:
No, not that I know off :-) I was having this function here to experiment with the loss function for negative examples, but I ended up removing that again. I double checked, and the current code is giving the exact same output as your original function so I'll remove this again (should have done that before submitting)